server/middleware/authentication.js

```diff
@@ -364,3 +381,28 @@ export function mustBeLoggedIn(req, res, next) {
     }
   });
 }
+
+export const checkTwoFactorAuthJWT = (req, res, next) => {
+  let token;
+  try {
+    token = getTokenFromRequestHeaders(req);
+  } catch (err) {
+    return next(err);
+  }
+
+  jwt.verify(token, jwtSecret, (err, decoded) => {
+    // JWT library either returns an error or the decoded version
+    if (err) {
+      return next(new BadRequest(err.message));
+    } else {
+      req.jwtPayload = decoded;
+    }
+  });
+
```

Comment:
I think there is a potential problem here, if `jwt.verify` is asynchronous then we should wait before moving to the next block. Currently, there is no guarantee that the `jwt.verify` is done.

Maybe by performing `jwt.verify` with async/await instead of a callback you'll see better the problem.